### DISCUSSION TITLE
Calculating Combinations Of Equipment

### DISCUSSION THREAD

**Discussion by F0RM4TION** [LukeCroteau/RaidExtractor#25] [Q&A] [answered]
so ive separated gear based on equipment type and now im stuck on how i would be able to get every kind of combination possible. anyone got any ideas? or know what kind of calculation the website does to get the combinations?

**Comment by andreas-henning** [comment, marked as answer]
Attempting to calculate every possible combination is futile. If you have just 60 artifacts of each type (boots, gauntlets, etc), that will already yield over 40 billion combinations. No brute force approach will work.

Wish I had a more constructive answer. I have no idea what HH's website does behind the scenes.

**Reply by F0RM4TION** [reply]
yea i realised it was impossible after abit of thinking. i took a look at HH's video on the optimizer and noticed that on the start optimizing page that he limits the search to the 6 best artifacts per slot. so thats how he probably cuts down the amount of calculation that needs to be done. only thing i dont know is how he is determining what the 6 best artifacts are for each slot.

https://youtu.be/ZoHoKgy4RfU?t=756

**Comment by kfishkin** [comment]
I think HH's video is misleading, I don't think he realizes that max(2) takes as long as max(200). It's about how many options you look at, not how many you keep. The biggest thing I notice that will reduce HH's search space over this one is that you can exclude types of armor (e.g. "no Destroy gear"), that will help a lot. 
One technique he mentions in the video is to freeze jewelry and do armor, and then do the reverse.

**Reply by andreas-henning** [reply]
Hmm... I would have claimed the exact opposite, but I'm not sure we're using the same terminology.

After determining which are the best artifacts of each type (boots, gauntlets, etc), it does look like HH's website then checks every possible combination and then displays the 10 best combinations.

If you KEEP the six best artifacts of each type, you're left with approximately 46000 combinations. That's not a big deal. If you KEEP the 10 best artifacts of each type, you're left with 10 million combinations... the number of combinations quickly spirals out of control, so how many you KEEP does make a big difference.

So how do we select the six "best" artifacts per type? I don't know, but maybe this...

If we're looking at boots, and the user specified they are interested in HP and Speed, you could create two lists of boots, one sorted by HP and one sorted by Speed. If HP and Speed were given equal weight, you'd then select the first six boots which appear in both lists. Different weighting would cause you to select more from one list than the other.

I don't know if that's how HH does it, but I suspect it would be something along those lines...

**Comment by F0RM4TION** [comment]
got a feeling that HH's is determining the 6 best artifacts based on the importance value in the previous step so maybe looking for the piece of gear that has the majority of the important stats as well as the highest rolls on those stats and assuming that they are more likely to give a better build and in the process weeding out alot of the other gear that would be pointless to check

**Reply by F0RM4TION** [reply]
thanks, i managed to do that calculation, but it took afew mins to process

**Comment by F0RM4TION** [comment]
Been doing some experimenting and i think ill settle on the best 4 pieces of equipment per slot. takes about 2 seconds to load and it should give a decent amount of accuracy for the best build as long as i find a good way of cheaply working out the best peices of gear for each slot

262,140 combinations total

**Comment by RolandKol** [comment]
I am not any kind of dev, but if script sort items by stats before combining them into bulks to get stats needed.
So the top ones would be processed first and the result could be shown a bit faster.
